sphinx-doc/howto/determine-the-most-efficient-device.rst

```diff
@@ -0,0 +1,144 @@
+     - 18168
+   * - CPU
+     - 32
+     - 22394
+   * - CPU
+     - 64
+     - 25031
+   * - GPU
+     - 1
+     - 15955
+
+The optimal device selection depends on the metric. When the metric is wall clock time only, choose
+the highest performance benchmark. When the metric is a cost, choose based on the efficiency of each
+device configuration:
+
+.. math::
+
+    \eta = \frac{S}{S_\mathrm{ref}} \cdot \frac{C_\mathrm{ref}}{C}
+
+where :math:`\eta` is the efficiency, :math:`S` is the performance :math:`C` is the cost and the
+:math:`\mathrm{ref}` subscript denotes the reference.
+
+One cost metric is compute time. Most HPC resources assign a cost by CPU core hours:
+
+.. math::
+
+    \frac{C_\mathrm{1\ CPU}}{C_\mathrm{P\ CPUs}} = \frac{1}{P}
+
+Some HPC resources may assign an effective cost to GPUs. When this is not the case, use the ratio of
+available GPU hours to CPU core hours as a substitute. This example will assign a relative cost of
+
+.. math::
+
+    \frac{C_\mathrm{1\ CPU}}{C_\mathrm{1\ GPU}} = \frac{1}{64}.
+
+.. image:: wca-efficiency-2048.svg
+    :alt: Performance and efficiency of 2048 particle WCA simulations.
+
+With 2048 particles in this example, the CPU is always more efficient than the GPU and the CPU is
+faster than the GPU when :math:`P \ge 16`. Therefore, the CPU is always the optimal choice. Choose a
+number of ranks :math:`P` depending on project needs and budgets. Larger values of :math:`P` will
+provide results with lower latency at the cost of more CPU core hours. In this example, :math:`P=8`
+(:math:`\eta \sim 0.6`) is a middle ground providing a significant reduction in time to solution at
+a moderate extra cost in CPU core hours.
+
+Example Results (N=131,072)
+---------------------------
+
+The results are very different with 131,072 particles
+(``$ mpirun -n $P python3 determine-the-most-efficient-device.py --device $PROCESSOR --replicate=4``):
+
+.. list-table::
+   :header-rows: 1
+
+   * - Processor
+     - P
+     - TPS
+   * - CPU
+     - 1
+     - 36.072
+   * - CPU
+     - 2
+     - 61.988
+   * - CPU
+     - 4
+     - 143.25
+   * - CPU
+     - 8
+     - 281.35
+   * - CPU
+     - 16
+     - 502.48
+   * - CPU
+     - 32
+     - 910.58
+   * - CPU
+     - 64
+     - 1451.5
+   * - CPU
+     - 128
+     - 2216.1
+   * - CPU
+     - 256
+     - 2706.8
+   * - GPU
+     - 1
+     - 7276.5
+
+.. image:: wca-efficiency-131072.svg
+    :alt: Performance and efficiency of 131,072 particle WCA simulations.
+
+At a this system size, the GPU is always both faster and more efficient than the CPU.
+
+Compare the two examples and notice that the TPS achieved by the GPU is only cut in half when the
+system size is increased by a factor of 64. This signals that the smaller system size was not able
+to utilize all the parallel processing units on the GPU.
+
+.. note::
+
+    Use trial moves per second (`hoomd.hpmc.integrate.HPMCIntegrator.mps`) as the performance
+    metric when benchmarking HPMC simulations.
```

Comment:
Would using trial moves per second give different results than using TPS? Doesn't an MC timestep involve a fixed number of trial moves per particle?

Reply:
Yes, trial moves per second will give different results. There are *not* a fixed number of trial moves per timestep - it varies with the number of MPI ranks and is also different on the GPU than the CPU.